File: Step_Response.m
clear all % Clears all defined MATLAB variables

u = 0.1; % Step PWM duty cycle value (between 0 and 1)
dt = 0.5; % Sampling interval in seconds
t_end = 10 * 60; % Time to record in seconds (converted from 10 minutes)
T_Safe = 120; % Safe temperature threshold in °F

% Create an Arduino object. Replace 'ComX' with the correct COM port (e.g., 'COM3')
a = arduino('ComX', 'Nano');

% Record the start time
startTime = datetime('now');

% Apply the initial PWM signal to the desired pin. Replace 'pinX' with the appropriate pin (e.g., 'D9')
writePWMDutyCycle(a, 'pinX', u);

% Initialize time tracking and data logging
Delta_t = seconds(datetime('now') - startTime); % Time elapsed in seconds
Count = 0; % Initialize counter for data points

% Main loop to record temperature until `t_end` is reached
while Delta_t < t_end
    % Read the current temperature in °F from the sensor connected to 'A0'
    PV = readVoltage(a, 'A0') * 180 + 32; % Adjust pin if necessary
    
    % Check if the temperature exceeds the safe threshold
    if PV > T_Safe
        writePWMDutyCycle(a, 'pinX', 0); % Stop the PWM signal if temperature is unsafe
        disp('Temperature exceeded safe level...Process stopped.')
        break; %Terminate the process
    end
    
    % Update the time elapsed
    Delta_t = seconds(datetime('now') - startTime);
    
    % Increment the counter and store data
    Count = Count + 1;
    t(Count) = Delta_t; % Store the time elapsed
    T(Count) = PV; % Store the temperature reading
    
    % Display the current time and temperature
    disp([t(Count), T(Count)]);
    
    % Pause for the sampling interval
    pause(dt);
end

% Plot the recorded temperature data
plot(t, T);
xlabel('Time (s)');
ylabel('Temperature (°F)');
title('Step Response');
grid on;

File: Step_Response.cpp
// BE 441/541 Instrumentation and Controls
// Arduino code to collect step response data 

#define pwmPin  3     // Replace with the appropriate pin
#define tempPin  A0   // Replace with the appropriate analog pin
const float u = 0.1;  // Step PWM duty cycle value (between 0 and 1)
const float dt = 0.5; // Sampling interval in seconds
const int t_end = 10 * 60; // Time to record in seconds (converted from 10 minutes)
const float T_Safe = 120; // Safe temperature threshold in °F

unsigned long startTime;
unsigned long currentTime;
unsigned long elapsedTime;
int count = 0;

void setup() {
  Serial.begin(9600);
  pinMode(pwmPin, OUTPUT);
  Serial.println("Time(s), Temperature(°F)");
  analogWrite(pwmPin, u * 255); // Apply the initial PWM signal
  startTime = millis(); // Record the start time
}

void loop() {
  currentTime = millis();
  elapsedTime = (currentTime - startTime) / 1000; // Time elapsed in seconds

  if (elapsedTime < t_end) {
    float voltage = analogRead(tempPin) * (5.0 / 1023.0); // Read the current voltage
    float PV = voltage * 180 + 32; // 10 mV/°C = 100°C/V. Convert temperature to °F

    if (PV > T_Safe) {
      analogWrite(pwmPin, 0); // Stop the PWM signal if temperature is unsafe
      Serial.println("Temperature exceeded safe level...Process stopped.");
      while (true); // Terminate the process
    }

    Serial.print(elapsedTime);
    Serial.print(", ");
    Serial.println(PV);
    delay(dt * 1000); // Pause for the sampling interval
  }
}

    delay(dt * 1000); // Pause for the sampling interval
  }
}
